storm-core/src/jvm/backtype/storm/scheduler/resource/strategies/eviction/DefaultEvictionStrategy.java

```diff
@@ -0,0 +1,109 @@
+import backtype.storm.scheduler.TopologyDetails;
+import backtype.storm.scheduler.WorkerSlot;
+import backtype.storm.scheduler.resource.RAS_Nodes;
+import backtype.storm.scheduler.resource.User;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Map;
+
+public class DefaultEvictionStrategy implements IEvictionStrategy {
+    private static final Logger LOG = LoggerFactory
+            .getLogger(DefaultEvictionStrategy.class);
+
+    private Topologies topologies;
+    private Cluster cluster;
+    private Map<String, User> userMap;
+    private RAS_Nodes nodes;
+
+    @Override
+    public void prepare(Topologies topologies, Cluster cluster, Map<String, User> userMap, RAS_Nodes nodes) {
+        this.topologies = topologies;
+        this.cluster = cluster;
+        this.userMap = userMap;
+        this.nodes = nodes;
+    }
+
+    @Override
+    public boolean makeSpaceForTopo(TopologyDetails td) {
+        LOG.debug("attempting to make space for topo {} from user {}", td.getName(), td.getTopologySubmitter());
+        User submitter = this.userMap.get(td.getTopologySubmitter());
+        if (submitter.getCPUResourceGuaranteed() == null || submitter.getMemoryResourceGuaranteed() == null) {
+            return false;
+        }
+        double cpuNeeded = td.getTotalRequestedCpu() / submitter.getCPUResourceGuaranteed();
+        double memoryNeeded = (td.getTotalRequestedMemOffHeap() + td.getTotalRequestedMemOnHeap()) / submitter.getMemoryResourceGuaranteed();
+
+        User evictUser = this.findUserWithHighestAverageResourceUtilAboveGuarantee();
+        //user has enough resource under his or her resource guarantee to schedule topology
+        if ((1.0 - submitter.getCPUResourcePoolUtilization()) >= cpuNeeded && (1.0 - submitter.getMemoryResourcePoolUtilization()) >= memoryNeeded) {
+            if (evictUser != null) {
+                TopologyDetails topologyEvict = evictUser.getRunningTopologyWithLowestPriority();
+                evictTopology(topologyEvict);
+                return true;
+            }
+        } else {
+            if (evictUser != null) {
+                if ((evictUser.getResourcePoolAverageUtilization() - 1.0) > (cpuNeeded + (submitter.getResourcePoolAverageUtilization() - 1.0))) {
+                    TopologyDetails topologyEvict = evictUser.getRunningTopologyWithLowestPriority();
+                    evictTopology(topologyEvict);
+                    return true;
+                }
+            }
+        }
+        //See if there is a lower priority topology that can be evicted from the current user
+        for (TopologyDetails topo : submitter.getTopologiesRunning()) {
+            //check to if there is a topology with a lower priority we can evict
+            if (topo.getTopologyPriority() > td.getTopologyPriority()) {
+                evictTopology(topo);
+                return true;
+            }
+        }
```

Comment:
I am wondering if it would be good to log the reason a topology is evicted along with the other details?

It might make it easier for an admin to explain what happened beyond that the topology was evicted.